File: Design-Patterns/Singleton/Cpp/Singleton/singleton.cpp
#include "singleton.h"

#include <iostream>

singleton::singleton()
{
	data_ = new int(10);
}

/*
The declaration of a static data member in its class definition is not a definition and may be of an incomplete type other than cv-qualified void.
The definition for a static data member shall appear in a namespace scope enclosing the member’s class definition.
Link: https://www.ibm.com/docs/en/zos/2.1.0?topic=members-static-data
 */
singleton* singleton::instance_ = nullptr;

void singleton::operation()
{
	std::cout << "Singleton" << std::endl;
}

void* singleton::get_data() const
{
	return data_;
}

singleton* singleton::get_instance()
{
	if (instance_ == nullptr)
	{
		instance_ = new singleton();
	}

	return instance_;
}

File: Design-Patterns/Singleton/CSharp/Singleton/Singleton.cs
namespace Singleton
{
    public class Singleton
    {
        public static Singleton Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new Singleton();
                }

                return _instance;
            }
        }

        private static Singleton _instance;

        private object _data;

        private Singleton()
        {
            _data = new object();
        }

        public void Operation()
        {

        }

        public object GetData()
        {
            return _data;
        }
    }
}
